drivers/serial/uart_nrfx_uarte.c

```diff
@@ -916,12 +987,34 @@ static void rxto_isr(const struct device *dev)
 static void txstopped_isr(const struct device *dev)
 {
 	struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 
 	if (!data->async->tx_buf) {
+		/* If there is a pending tx request, it means that uart_tx()
+		 * was called when there was ongoing uart_poll_out. Handling
+		 * TXSTOPPED interrupt means that uart_poll_out has completed.
+		 */
+		if (data->async->pend_tx_buf) {
+			int key = irq_lock();
+
+			if (nrf_uarte_event_check(uarte,
+						NRF_UARTE_EVENT_TXSTOPPED)) {
+				data->async->tx_buf = data->async->pend_tx_buf;
+				data->async->pend_tx_buf = NULL;
+				data->async->tx_amount = -1;
+				tx_start(uarte, data->async->tx_buf,
+					 data->async->tx_size);
+			}
+
+			irq_unlock(key);
+		}
 		return;
 	}
 
-	size_t amount = nrf_uarte_tx_amount_get(get_uarte_instance(dev));
+	k_timer_stop(&data->async->tx_timeout_timer);
+
+	size_t amount = (data->async->tx_amount >= 0) ?
+			data->async->tx_amount : nrf_uarte_tx_amount_get(uarte);
```

Comment:
You need to lock interrupts for these operations, as `uarte_nrfx_poll_out()` might potentially get called in between.

Reply:
good point, done.